### PR TITLE
consensus: add tx_root to block header

### LINKED ISSUE #847: node: Add transactions merkle tree root to the block header
**Describe the bug**
This is needed in order to reflect protocol changes made in https://github.com/dusk-network/dusk-blockchain/issues/1499

### PR DESCRIPTION
**Pull request by herr-seppia**
### `node-data`
- Add `txroot` to block header

### `consensus`

- Add `dusk-merkle` dependency
- Add `fn consensus::merkle::merkle_root` wrapping `dusk-merkle` to guarantee compatibility with dusk-blockchain's merkle-tree
- Add block header's `txroot` creation/verification

Resolves #847

### CODE REVIEW

**Review by moCello — COMMENTED** [review]
There might be a nicer way to get a 'smaller' merkle tree without needing to create one for each height and batch inserting leaves into it.

If you create a tree, insert leaves and get an opening, you can traverse the opening to the first node that has more than one children different from the `EMPTY_SUBTREES` of the height. The first note that has more than one children set is the root of the smallest tree that can hold all the leaves that have been inserted. If at each level there is only one child, the tree only holds one leaf and the parent of that leaf is the smallest root

**Comment by herr-seppia** [comment]
@moCello you're right

Pushed the implementation you suggested.
Feel free to have a look, in the meanwhile I want to sketch an additional approach

File: consensus/src/merkle.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    /// Returns the root of the smallest inner tree
+    fn shrunken_root(&self) -> [u8; 32] {
+        if self.tree.is_empty() {
+            return EMPTY_NODE.0;
+        }
+
+        let inner_tree_height =
+            (self.tree.len() as f64 - 1f64).sqrt().floor() as usize + 1;
```
Comment (moCello) [inline consensus/src/merkle.rs:79]: sqrt works only with ARITY=2 which is no problem in this case but not so nice for a general approach

File: consensus/src/merkle.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+        if inner_tree_height == H {
+            return self.tree.root().0;
+        }
+
+        let opening = self
+            .tree
+            .opening(0)
```
Comment (moCello) [inline consensus/src/merkle.rs:87]: We insert leaves at a specific position so there could be a case where there is no leaf at pos=0 but the tree is still not empty. Use the walker to retrieve any leaf instead

File: consensus/src/merkle.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            .expect("Opening to exists if tree is not empty");
+
+        opening.branch()[H - inner_tree_height - 1][0].0
```
Comment (moCello) [inline consensus/src/merkle.rs:90]: This also assumes that the leaves are all inserted in order, but the tree is sparse. I'm afraid you will need to traverse the branch to check for the first node that holds more than one children.

**Review by moCello — CHANGES_REQUESTED** [review]
Some change requests, see comments

**Review by moCello — APPROVED** [review]
Previous comments only apply to a general approach but in this special case it is actually not needed.
So LGTM

**Review by goshawk-3 — APPROVED** [review]

**Pull request merged**